test/groovy/DeployTest.groovy

```diff
@@ -0,0 +1,333 @@
+        ]
+    }
+
+    @Test
+    void errorNoTargetsDefined() {
+
+        nullScript.commonPipelineEnvironment.configuration = [
+            general: [
+                neoTargets: [],
+                cfTargets: []
+            ]
+        ]
+
+        thrown.expect(Exception)
+        thrown.expectMessage('Test Deployment skipped because no targets defined!')
+
+        stepRule.step.deploy(
+            script: nullScript
+        )
+    }
+
+    @Test
+    void neoDeploymentTest() {
+
+        nullScript.commonPipelineEnvironment.configuration.general.cfTargets = []
+
+        stepRule.step.deploy(
+            script: nullScript,
+            utils: utilsMock,
+            stage: 'test',
+            source: 'file.mtar'
+        )
+
+        Assert.assertThat(shellRule.shell,
+            new CommandLineMatcher().hasProlog('neo.sh deploy-mta')
+                .hasOption('synchronous', '')
+                .hasSingleQuotedOption('host', 'test\\.deploy\\.host1\\.com')
+                .hasSingleQuotedOption('account', 'trialuser1')
+                .hasSingleQuotedOption('user', 'user1')
+                .hasSingleQuotedOption('password', 'password1')
+                .hasSingleQuotedOption('source', 'file.mtar'))
+
+        Assert.assertThat(shellRule.shell,
+            new CommandLineMatcher().hasProlog('neo.sh deploy-mta')
+                .hasOption('synchronous', '')
+                .hasSingleQuotedOption('host', 'test\\.deploy\\.host2\\.com')
+                .hasSingleQuotedOption('account', 'trialuser2')
+                .hasSingleQuotedOption('user', 'user2')
+                .hasSingleQuotedOption('password', 'password2')
+                .hasSingleQuotedOption('source', 'file.mtar'))
+    }
+
+    @Test
+    void neoRollingUpdateTest() {
+
+        stepRule.step.deploy(
+            script: nullScript,
+            utils: utilsMock,
+            jenkinsUtils: jenkinsUtilsMock,
+            stage: 'test',
+            neoTargets: [
+                [
+                    host: 'test.param.deploy.host.com',
+                    account: 'trialparamNeoUser',
+                    credentialsId: 'paramNeoCredentialsId'
+                ]
+            ],
+            source: 'file.mtar',
+            enableZeroDowntimeDeployment: true 
+        )
+
+        Assert.assertThat(shellRule.shell,
+            new CommandLineMatcher().hasProlog('neo.sh deploy-mta')
+                .hasOption('synchronous', '')
+                .hasSingleQuotedOption('host', 'test\\.param\\.deploy\\.host\\.com')
+                .hasSingleQuotedOption('account', 'trialparamNeoUser')
+                .hasSingleQuotedOption('user', 'paramNeoUser')
+                .hasSingleQuotedOption('password', 'paramNeoPassword')
+                .hasSingleQuotedOption('source', 'file.mtar'))
+
+        assertThat(shellRule.shell, hasItem(containsString('cf login -u "cfUser1" -p \'cfPassword1\' -a https://api.cf.eu10.hana.ondemand.com -o "testOrg1" -s "testSpace1"')))
```

Comment:
The approach here is to have a step which traverses the configuration and prepares the more specific steps `neoDeploy` and `cfDeploy`. Here in all the tests we check the sh statements triggered by these more specific checks. This means the more specific steps are executed.

I would like to propose urgently to set the mock barrier here when the more specific steps are called (e.g. `registerAllowedMethod(`neoDeploy`, [Map], ...)`.

The disadvantage with the current approach of re-testing the more specific step is: in case we need to change something in the more specific steps, we need so adjust not only the tests for the more specific steps, but also this test here.

For testing the more specific steps we have already `CloudFoundryDeployTest` and `NeoDeployTest`. There is basically no need for testing the more specific steps here.

Reply:
Ok

Reply:
>it is testing the logic of the new step, but asserting the output of the specific steps [...]

I see we agree on what we would like to agree (that's cool): we would like to test the new step, only the new step. But we do not fully agree on how that should be accomplished.

Currently we execute also `cfDeploy` and `neoDeploy` when testing the new `deploy` step and assert the shell output of these steps. This means we have more code under test scope than we actually need. Would be sufficient to stop earlier when `cfDeploy` and `neoDeploy` are called and check that these steps are called with the expected parameters. For `cfDeploy` and `neoDeploy` we have already test cases.

Advantage: In case we have to change e.g. neoDeploy (...lets assume it needs a new flag `--foo`): When we have the code as it is currently in this PR we have to adjust `NeoDeployTest`  and this test here since the new flag `--foo` affects the `sh` output for both tests. But with the class under test for this test class here there is no change at all.

Why checking for `sh` output when I can also `registerAllowedMethod(neoDeploy, ...)` and check on this level? Looks to me like checking for bullet based on its shadows. Could also be cylinder in case we check only the [shadow at the bottom](http://www.chemgapedia.de/vsengine/media/vsc/de/ph/14/ep/einfuehrung/wellenoptik/bilder/sk0_2.png).

>  can change the assertions, but the rest stays the same.

I still vote for asserting at the level of `cfDeploy` and `neoDeploy` rather than checking the `sh` output of these steps.

Reply:
This was already clear .. To explain two times the same is not needed